chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,308 @@
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	// peers tracks the current online status of peers based on online/offline
+	// events.
+	peers map[[33]byte]bool
+
+	quit chan struct{}
+
+	sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// Get open channels provides a list of existing open channels
+	// which is used to populate the ChannelEventStore with a set of channels
+	// on startup.
+	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
+}
+
+// NewChannelEventStore initializes an event store with the config provided. Note
+// that this function does not start the monitoring process, Start() must be
+// called.
+func NewChannelEventStore(config *Config) *ChannelEventStore {
+	store := &ChannelEventStore{
+		cfg:      config,
+		channels: make(map[uint64]*chanEventLog),
+		peers:    make(map[[33]byte]bool),
+		quit:     make(chan struct{}),
+	}
+
+	return store
+}
+
+// Start adds all existing open channels to the event store and starts monitoring
+// for channel opened/closed events so that it can add and remove channels as
+// appropriate.
+func (c *ChannelEventStore) Start() error {
+	// Create a subscription to channel events.
+	channelClient, err := c.cfg.ChannelEvents()
+	if err != nil {
+		return err
+	}
+
+	// Create a subscription to peer events.
+	peerClient, err := c.cfg.PeerEvents()
+	if err != nil {
+		return err
+	}
+
+	// All events that need to be monitored are piped into this channel.
+	// This is done to assist mocking for testing.
+	events := make(chan interface{})
+
+	// Consume all events piped into the events channel.
+	c.wg.Add(1)
+	go c.monitorChannelEvents(events)
+
+	// This goroutine pipes all events into a single events channel, closing the
+	// channel and exiting if one of the event subscriptions fails, or the store
+	// is given the instruction to exit.
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+
+		// Cancel subscriptions on exit.
+		defer channelClient.Cancel()
+		defer peerClient.Cancel()
+
+		for {
+			select {
+			case e := <-channelClient.Updates():
+				events <- e
```

Comment:
should select on quit also

Reply:
Which quit, got both subscriptions and the store itself below?

Reply:
`c.quit` should be enough.